From 9c4db5da2a556ace646259520a8ce042f99423f2 Mon Sep 17 00:00:00 2001
From: Ahmad Abbas <ahmada@marvell.com>
Date: Sun, 9 Aug 2015 18:34:01 +0300
Subject: [PATCH 0323/1240] env: added initial support for NAND in runtime
 environment selection

added initial support structure for booting from nand when working on
runtime environment selection. the interface has been tested on a39x,
booting from spi while taking environment from nand flash.

Change-Id: I9e7104bd64e2749d25d9d0593d66ea7907186952
Signed-off-by: Ahmad Abbas <ahmada@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/22653
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Yehuda Yitschak <yehuday@marvell.com>
Reviewed-by: Hanna Hawa <hannah@marvell.com>
---
 common/env_nand.c              | 29 +++++++++++++++++++++++++++++
 include/configs/armada38x.h    |  2 +-
 include/configs/mvebu-common.h |  5 +++--
 include/env_bootdev.h          |  1 +
 4 files changed, 34 insertions(+), 3 deletions(-)

diff --git a/common/env_nand.c b/common/env_nand.c
index 9c9bb82..fea3029 100644
--- a/common/env_nand.c
+++ b/common/env_nand.c
@@ -38,7 +38,9 @@
 #define CONFIG_ENV_RANGE	CONFIG_ENV_SIZE
 #endif
 
+#ifndef CONFIG_ENV_IS_IN_BOOTDEV
 char *env_name_spec = "NAND";
+#endif
 
 #if defined(ENV_IS_EMBEDDED)
 env_t *env_ptr = &environment;
@@ -62,7 +64,11 @@ DECLARE_GLOBAL_DATA_PTR;
  * This way the SPL loads not only the U-Boot image from NAND but
  * also the environment.
  */
+#ifdef CONFIG_ENV_IS_IN_BOOTDEV
+static int _nand_env_init(void)
+#else
 int env_init(void)
+#endif
 {
 #if defined(ENV_IS_EMBEDDED) || defined(CONFIG_NAND_ENV_DST)
 	int crc1_ok = 0, crc2_ok = 0;
@@ -177,7 +183,11 @@ static int erase_and_write_env(const struct env_location *location,
 static unsigned char env_flags;
 #endif
 
+#ifdef CONFIG_ENV_IS_IN_BOOTDEV
+static int nand_saveenv(void)
+#else
 int saveenv(void)
+#endif
 {
 	int	ret = 0;
 	ALLOC_CACHE_ALIGN_BUFFER(env_t, env_new, 1);
@@ -300,7 +310,12 @@ int get_nand_env_oob(nand_info_t *nand, unsigned long *result)
 #endif
 
 #ifdef CONFIG_ENV_OFFSET_REDUND
+
+#ifdef CONFIG_ENV_IS_IN_BOOTDEV
+static void nand_env_relocate_spec(void)
+#else
 void env_relocate_spec(void)
+#endif
 {
 #if !defined(ENV_IS_EMBEDDED)
 	int read1_fail = 0, read2_fail = 0;
@@ -372,7 +387,12 @@ done:
  * device i.e., nand_dev_desc + 0. This is also the behaviour using
  * the new NAND code.
  */
+
+#ifdef CONFIG_ENV_IS_IN_BOOTDEV
+static void nand_env_relocate_spec(void)
+#else
 void env_relocate_spec(void)
+#endif
 {
 #if !defined(ENV_IS_EMBEDDED)
 	int ret;
@@ -402,3 +422,12 @@ void env_relocate_spec(void)
 #endif /* ! ENV_IS_EMBEDDED */
 }
 #endif /* CONFIG_ENV_OFFSET_REDUND */
+
+#ifdef CONFIG_ENV_IS_IN_BOOTDEV
+void nand_env_init(void)
+{
+	gd->arch.env_func.save_env = nand_saveenv;
+	gd->arch.env_func.init_env = _nand_env_init;
+	gd->arch.env_func.reloc_env = nand_env_relocate_spec;
+}
+#endif
diff --git a/include/configs/armada38x.h b/include/configs/armada38x.h
index 849aa46..e367234 100644
--- a/include/configs/armada38x.h
+++ b/include/configs/armada38x.h
@@ -64,7 +64,7 @@
 */
 
 /* Define NAND chip parameters */
-#ifdef CONFIG_MVEBU_NAND_BOOT
+#if defined(CONFIG_MVEBU_NAND_BOOT) || defined(CONFIG_ENV_IS_IN_BOOTDEV)
 #define CONFIG_MVEBU_NAND_PAGE_SIZE	4096
 #define CONFIG_MVEBU_NAND_BLOCK_SIZE	128
 #define CONFIG_MVEBU_NAND_CELL_TYPE	MLC
diff --git a/include/configs/mvebu-common.h b/include/configs/mvebu-common.h
index ee02a6e..68a79e5 100644
--- a/include/configs/mvebu-common.h
+++ b/include/configs/mvebu-common.h
@@ -147,6 +147,7 @@
 
 #if defined(CONFIG_ENV_IS_IN_BOOTDEV)
 	#define CONFIG_ENV_IS_IN_SPI_FLASH
+	#define CONFIG_ENV_IS_IN_NAND
 	#define DEFAULT_BUBT_DST "spi"
 	#define CONFIG_DOIMAGE_TYPE     "flash"
 	#define CONFIG_DOIMAGE_SUFFIX   "dynamic"
@@ -176,14 +177,14 @@
 
 
 /* Boot from NAND settings */
-#if defined(CONFIG_MVEBU_NAND_BOOT)
+#if defined(CONFIG_MVEBU_NAND_BOOT) && !defined(CONFIG_ENV_IS_IN_BOOTDEV)
 	#define CONFIG_ENV_IS_IN_NAND
 	#define CONFIG_DOIMAGE_TYPE	"nand"
 	#define CONFIG_DOIMAGE_SUFFIX	"nand"
 #endif /* CONFIG_MVEBU_NAND_BOOT */
 
 /* Boot from SPI settings */
-#if defined(CONFIG_MVEBU_SPI_BOOT)
+#if defined(CONFIG_MVEBU_SPI_BOOT) && !defined(CONFIG_ENV_IS_IN_BOOTDEV)
 	#define CONFIG_ENV_IS_IN_SPI_FLASH
 	#define CONFIG_DOIMAGE_TYPE	"flash"
 	#define CONFIG_DOIMAGE_SUFFIX	"spi"
diff --git a/include/env_bootdev.h b/include/env_bootdev.h
index 433b939..7c51778 100644
--- a/include/env_bootdev.h
+++ b/include/env_bootdev.h
@@ -31,5 +31,6 @@ struct env_func_ptr {
 };
 
 void sf_env_init(void);
+void nand_env_init(void);
 
 #endif /* _ENV_BOOTDEV_H_ */
-- 
1.9.1

